ERCS/erc-7786.md

```diff
@@ -0,0 +1,222 @@
+```
+
+##### `setMessageExecuted`
+
+Checks that there is a valid and as yet unexecuted message whose contents are exactly those passed as arguments and whose receiver is the caller of the function. The message key MAY be an identifier, or another piece of data necessary for validation.
+
+MUST revert if the message is invalid or has already been executed.
+
+TBD: Passing full payload or payload hash (as done by Axelar). Same question for attributes, possibly different answer depending on attribute malleability.
+
+#### Dual Active-Passive Mode
+
+A gateway MAY operate in both active and passive modes, or it MAY switch from operating exclusively in active mode to passive mode or vice versa.
+
+A receiver SHOULD support both active and passive modes for any gateway. This is accomplished by first checking whether the caller of `receiveMessage` is a known gateway, and only validating the message if it is not; the first case supports an active mode gateway, while the second case supports a passive mode gateway.
+
+### Properties
+
+The protocol underlying a pair of gateways is expected to guarantee a series of properties. For detailed definition and discussion we refer to XChain Research’s [Cross-chain Interoperability Report](https://github.com/0xFableOrg/xchain/blob/7789933bba24e2dd893cf515157af70474e7180b/README.md).
+
+- The protocol MUST guarantee Safety: A message is delivered at the destination if and only if it was sent at the source, and the sending transaction was finalized.
+- The protocol MUST guarantee Liveness: A sent message is delivered at the destination eventually, assuming Liveness and censorship-resistance of the source and destination chains.
+- The protocol SHOULD guarantee Timeliness: A sent message is delivered at the destination within a bounded delivery time, which should be documented.
+- The above properties SHOULD NOT rely on trust in some centralized actor. For example, safety should be guaranteed by some trustless mechanism such as a light client proof, or attestations by an open, decentralized validator set. Relaying should be decentralized or permissionless to ensure liveness; a centralized relayer can fail and thus halt the protocol.
+
+### Pending discussion
+
+- How to "reply" to a message? Duplex gateway? Getter for reverse gateway address? Necessary for some applications, e.g., recovery from token bridging failure?
+- Should the destination and receiver inputs of `sendMessage` be kept as two separate strings, or merged as a single CAIP-10 string with a `:` separator? This has implication of the calldata length, which in some cases may be stored.
+- Do we want the gateway to have the ability to inform users of the address of the new version, similar to how `AccessManager` can update then authority trusted by an `AccessManaged`? This could be useful if a gateway is ever deprecated in favor of a new version.
+- Should data and metadata attributes be split in two? What are data attributes used for? Do we need them?
```

Comment:
I would think that in the ERC-20 use case, it would be metadata, the gateway would look at it and either transfer the ERC-20 before calling the receiver or (arguably better), set an allowance to allow the destination to transfer the tokens to itself.

In this case the metadata can be stripped, or passed on to the receiver. (It stripped, might require the receiver to encode it in its own data format in some cases. In other cases, it's not needed (e.g. buying an NFT whose price is fixed — in that case it will try to transfer the requested amount and either succeed or fail).

Data on the other hand, seems unnecessary, since the sender already receives a payload it can use to encode whatever it needs. If it's useless to the gateways, it should be in the payload.

Reply:
Btw, the ERC-20 use case is a nice illustration of the usefulness of metadata . A single-destination messaging system will really struggle to transfer ERC-20 and do a contract call at once (it would require something like making sure the ERC-20 contract itself be able to forward arbitrary calls to contracts).

Now, for that use case, a possibly simpler solution would be to enable multicalls in the messaging interface. This needs to be at the gateway level because simply deploying a "multicall receiver" would not work: the caller of the ERC-20 contract to mint a token would be the multicall receiver instead of the gateway.

I think it would be great to list off more potential use cases for metadata properties, but it seems potentially useful enough (for things we may not even foresee) that I would definitely keep it.

I think the multicall system is probably very generally useful as well, and you may want to consider including it. EIP-4337 (in which strangely I see many parallels to this despite their very different nature) has a similar capability as well.

Reply:
>If it's useless to the gateways, it should be in the payload.

I like this formulation! I actually think it's kind of implied by the fact that gateways have to support the attributes you're including, otherwise they'll reject the message. 

Regarding multicalls, I'd reframe it as "atomic multi-messages" or something like that. I'm a bit reluctant given that most protocols I've seen don't support this (I only recall LayerZero having this). Additionally it could technically be implemented as an attribute. Definitely something to consider though.